snuba/cli/multistorage_consumer.py

```diff
@@ -0,0 +1,200 @@
+    topic = Topic(topics.pop())
+    if topics:
+        raise ValueError("only one topic is supported")
+
+    # XXX: The ``CommitLogConsumer`` also only supports a single topic at this
+    # time. (It is less easily modified.) This also assumes the commit log
+    # topic is on the same Kafka cluster as the input topic.
+    commit_log_topics = {
+        spec.topic_name
+        for spec in (
+            storage.get_table_writer().get_stream_loader().get_commit_log_topic_spec()
+            for storage in storages.values()
+        )
+        if spec is not None
+    }
+
+    commit_log_topic: Optional[Topic]
+    if commit_log_topics:
+        commit_log_topic = Topic(commit_log_topics.pop())
+    else:
+        commit_log_topic = None
+
+    if commit_log_topics:
+        raise ValueError("only one commit log topic is supported")
+
+    # XXX: This requires that all storages are associated with the same Kafka
+    # cluster so that they can be consumed by the same consumer instance.
+    # Unfortunately, we don't have the concept of independently configurable
+    # Kafka clusters in settings, only consumer configurations that are
+    # associated with storages and/or global default configurations. To avoid
+    # implementing yet another method of configuring Kafka clusters, this just
+    # piggybacks on the existing configuration method(s), with the assumption
+    # that most deployments are going to be using the default configuration.
+    storage_keys = [*storages.keys()]
+    consumer_configuration = build_kafka_consumer_configuration(
+        storage_keys[0],
+        consumer_group,
+        auto_offset_reset=auto_offset_reset,
+        queued_max_messages_kbytes=queued_max_messages_kbytes,
+        queued_min_messages=queued_min_messages,
+    )
+
+    for storage_key in storage_keys[1:]:
+        if (
+            build_kafka_consumer_configuration(storage_key, consumer_group)[
+                "bootstrap.servers"
+            ]
+            != consumer_configuration["bootstrap.servers"]
+        ):
+            raise ValueError("storages cannot be located on different Kafka clusters")
+
+    if commit_log_topic is None:
+        consumer = KafkaConsumer(consumer_configuration)
+    else:
+        # XXX: This relies on the assumptions that a.) the Kafka cluster where
+        # the commit log topic is located is the same as the input topic (there
+        # is no way to specify otherwise, at writing) and b.) all storages are
+        # located on the same Kafka cluster (validated above.)
+        producer = ConfluentKafkaProducer(
+            build_kafka_producer_configuration(storage_keys[0])
+        )
+        consumer = KafkaConsumerWithCommitLog(
+            consumer_configuration,
+            producer=producer,
+            commit_log_topic=commit_log_topic,
+        )
+
+    metrics = MetricsWrapper(environment.metrics, "consumer")
+    processor = StreamProcessor(
```

Comment:
It should not be too hard to perform the actual act of message filtering since it only requires wrapping the processor (as of #1408, [example](https://github.com/getsentry/snuba/blob/f13141774676f64d376d76805f13b5fd0977d74d/snuba/consumers/consumer_builder.py#L222-L224).) I also do not think it would be particularly technically difficult to get that snapshot state into `process_message_multistorage` to be able to perform the necessary filtering (the state data should be trivially serializable.) I do not have a plan for how that state is provided to the consumer itself for each storage independently without re-familiarizing myself with the snapshot consumer.